File: samples/system/python/06_cost_optimization.py
"""
06 - Cost Optimization with Budget Limits
==========================================

Cost-first selection strategy combined with per-provider budget limits.
ModelMesh routes each request to the cheapest active model and automatically
deactivates providers when their budget is exhausted.

This sample demonstrates:
  - Configuring cost-first rotation policy
  - How the system routes to the cheapest models first
  - Budget-based deactivation when spend limits are reached
  - Multiple providers with different cost characteristics

Uses mock providers so it runs without API keys.
"""

import asyncio
from modelmesh import ModelMesh, MeshConfig
from modelmesh.interfaces.provider import (
    ChatMessage, CompletionChoice, CompletionRequest, CompletionResponse,
    ErrorClassification, ModelInfo, ModelPricing, ProviderConnector,
    QuotaStatus, RateLimitStatus, TokenUsage,
)


class MockCostProvider(ProviderConnector):
    """Mock provider with cost tracking."""

    def __init__(self, name: str, cost_per_request: float = 0.01):
        self._name = name
        self._cost = cost_per_request
        self._total_cost = 0.0
        self._requests = 0

    async def complete(self, request: CompletionRequest) -> CompletionResponse:
        self._requests += 1
        self._total_cost += self._cost
        return CompletionResponse(
            id=f"resp-{self._name}-{self._requests}",
            model=self._name,
            choices=[CompletionChoice(
                index=0,
                message=ChatMessage(
                    role="assistant",
                    content=f"[{self._name}] Response #{self._requests} (cost: ${self._cost:.4f})",
                ),
                finish_reason="stop",
            )],
            usage=TokenUsage(prompt_tokens=10, completion_tokens=15, total_tokens=25),
        )

    async def stream(self, request):
        yield await self.complete(request)

    def get_capabilities(self):
        return ["generation.text-generation.chat-completion"]

    def supports(self, cap):
        return cap in self.get_capabilities()

    def list_models(self):
        return [ModelInfo(id=self._name, name=self._name)]

    def get_model_info(self, mid):
        return ModelInfo(id=mid, name=mid)

    def check_quota(self):
        return QuotaStatus(used=self._requests)

    def get_rate_limits(self):
        return RateLimitStatus()

    def get_pricing(self, mid):
        return ModelPricing()

    def report_usage(self, mid, usage):
        pass

    def classify_error(self, error):
        return ErrorClassification(retryable=False)


async def main() -> None:
    # Create providers with different cost levels
    prov_cheap = MockCostProvider("deepseek-chat", cost_per_request=0.001)
    prov_free = MockCostProvider("llama-3-groq", cost_per_request=0.0)
    prov_premium = MockCostProvider("gpt-4o", cost_per_request=0.03)

    config = MeshConfig(raw={
        "providers": {
            "deepseek": {"connector": "deepseek", "enabled": True, "instance": prov_cheap},
            "groq": {"connector": "groq", "enabled": True, "instance": prov_free},
            "openai": {"connector": "openai", "enabled": True, "instance": prov_premium},
        },
        "models": {
            "deepseek.chat": {
                "provider": "deepseek",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
            "groq.llama-3": {
                "provider": "groq",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
            "openai.gpt-4o": {
                "provider": "openai",
                "capabilities": ["generation.text-generation.chat-completion"],
            },
        },
        "pools": {
            "text-generation": {
                "capability": "generation.text-generation.chat-completion",
                "strategy": "cost-first",
            },
        },
    })
    mesh = ModelMesh()
    mesh.initialize(config)
    client = mesh.get_client()

    print("=" * 60)
    print("Cost Optimization with Budget Limits")
    print("=" * 60)
    print()
    print("DeepSeek ($0.001/req) < Groq (free) < GPT-4o ($0.030/req)")
    print("Cost-first strategy routes to the cheapest active model.\n")

    prompts = [
        "What is machine learning?",
        "Explain gradient descent.",
        "What is a neural network?",
        "Define overfitting.",
        "What is regularization?",
    ]

    for i, prompt in enumerate(prompts, start=1):
        response = client.chat.completions.create(
            model="text-generation",
            messages=[{"role": "user", "content": prompt}],
            max_tokens=100,
        )
        print(f"  Request {i}: {response.choices[0].message.content}")

    print(f"\n--- Cost Summary ---")
    print(f"  DeepSeek : {prov_cheap._requests} requests, ${prov_cheap._total_cost:.4f}")
    print(f"  Groq     : {prov_free._requests} requests, ${prov_free._total_cost:.4f}")
    print(f"  OpenAI   : {prov_premium._requests} requests, ${prov_premium._total_cost:.4f}")
    total = prov_cheap._total_cost + prov_free._total_cost + prov_premium._total_cost
    print(f"  Total    : ${total:.4f}")

    mesh.shutdown()
    print("\nModelMesh shut down.")


if __name__ == "__main__":
    asyncio.run(main())

File: samples/system/typescript/06_costOptimization.ts
/**
 * 06 - Cost Optimization with Budget Limits
 *
 * Configures a cost-first selection policy with daily and monthly budget
 * limits per provider. This sample demonstrates:
 *
 *   - Cost-first rotation policy: cheapest active model is always selected
 *   - Per-provider budget limits (daily and monthly)
 *   - Budget-based deactivation: when a provider's spend cap is reached,
 *     its models move to standby and the next cheapest provider takes over
 *   - Multiple providers with different cost profiles
 *
 * This pattern is ideal for applications that need to minimize spend while
 * maintaining service availability. The system exhausts cheap providers
 * before falling back to more expensive ones.
 *
 * Prerequisites:
 *   - Set DEEPSEEK_API_KEY, GROQ_API_KEY, OPENAI_API_KEY, and
 *     ANTHROPIC_API_KEY environment variables
 */

import { ModelMesh, MeshConfig } from "modelmesh-lite";

async function main(): Promise<void> {
  // -----------------------------------------------------------------------
  // 1. Configure providers ordered by cost (cheapest first)
  // -----------------------------------------------------------------------
  // DeepSeek and Groq offer the lowest per-token pricing. OpenAI and
  // Anthropic are more expensive but have higher rate limits and broader
  // capability support. The cost-first policy routes to the cheapest
  // active model regardless of declaration order.
  const config: MeshConfig = {
    raw: {
      providers: {
        // DeepSeek: ultra-low-cost chat and reasoning
        "deepseek.llm.v1": {
          enabled: true,
          api_key: "${secrets:DEEPSEEK_API_KEY}",
          budget: {
            daily_limit: 0.50,
            monthly_limit: 5.00,
          },
        },

        // Groq: fast free-tier inference
        "groq.inference.v1": {
          enabled: true,
          api_key: "${secrets:GROQ_API_KEY}",
          budget: {
            daily_limit: 0.00, // Free tier only
            monthly_limit: 0.00,
          },
        },

        // OpenAI: moderate cost, broad capabilities
        "openai.llm.v1": {
          enabled: true,
          api_key: "${secrets:OPENAI_API_KEY}",
          budget: {
            daily_limit: 2.00,
            monthly_limit: 20.00,
          },
        },

        // Anthropic: premium tier, strongest reasoning
        "anthropic.llm.v1": {
          enabled: true,
          api_key: "${secrets:ANTHROPIC_API_KEY}",
          budget: {
            daily_limit: 5.00,
            monthly_limit: 50.00,
          },
        },
      },

      pools: {
        "text-generation": {
          // Cost-first selects the cheapest active model for each request.
          // Provider pricing metadata is used to rank models by cost per
          // token. When a provider's budget limit is reached, its models
          // move to standby and the next cheapest active model takes over.
          strategy: "modelmesh.cost-first.v1",

          deactivation: {
            retry_limit: 3,
            error_codes: [429, 500, 503],
            // Budget-based deactivation: when cumulative spend reaches
            // the budget_limit, the model is deactivated for the period.
            budget_limit: 2.00,
          },

          recovery: {
            cooldown: "60s",
            on_quota_reset: true,
            quota_reset_schedule: "daily_utc",
          },

          retry: {
            max_attempts: 2,
            backoff: "exponential_jitter",
            initial_delay: "500ms",
            retryable_codes: [429, 500, 502, 503],
            scope: "same_provider",
          },
        },
      },

      observability: {
        routing: {
          connector: "modelmesh.console.v1",
        },
        // Track cost statistics to a file for analysis
        statistics: {
          connector: "modelmesh.local-file.v1",
          path: "./cost-stats.json",
          flush_interval: "30s",
        },
      },
    },
  };

  // -----------------------------------------------------------------------
  // 2. Initialize
  // -----------------------------------------------------------------------
  const mesh = new ModelMesh();
  await mesh.initialize(config);
  console.log("ModelMesh initialized with cost-first routing.");
  console.log("Providers (cheapest first): DeepSeek, Groq, OpenAI, Anthropic\n");

  const client = mesh.getClient();

  // -----------------------------------------------------------------------
  // 3. Send requests and observe cost-driven routing
  // -----------------------------------------------------------------------
  const prompts = [
    "Summarize the benefits of renewable energy in 2 sentences.",
    "What are the main programming paradigms?",
    "Explain the difference between TCP and UDP.",
    "Name three famous mathematicians and their contributions.",
    "What is the time complexity of quicksort?",
  ];

  console.log("--- Sending Requests (Cost-First Routing) ---\n");

  for (let i = 0; i < prompts.length; i++) {
    const response = await client.chat.completions.create({
      model: "text-generation",
      messages: [{ role: "user", content: prompts[i] }],
      max_tokens: 100,
    });

    console.log(`[${i + 1}] Prompt : ${prompts[i].substring(0, 50)}...`);
    console.log(`    Model : ${response.model}`);
    console.log(`    Tokens: in=${response.usage.prompt_tokens}, out=${response.usage.completion_tokens}`);
    console.log(`    Answer: ${(response.choices[0] as any).message.content.substring(0, 80)}...\n`);
  }

  // -----------------------------------------------------------------------
  // 4. Check cost statistics
  // -----------------------------------------------------------------------
  console.log("--- Cost Summary ---");
  // The stats() API provides per-provider cost totals.
  // In a real application, you would query mesh.stats() here.
  // For this sample, the statistics are flushed to cost-stats.json
  // by the observability connector.
  const pools = mesh.listPools();
  console.log(`Active pools: ${pools.length}`);
  console.log("Cost data written to ./cost-stats.json");

  // -----------------------------------------------------------------------
  // 5. Shut down
  // -----------------------------------------------------------------------
  await mesh.shutdown();
  console.log("\nModelMesh shut down.");
}

main().catch(console.error);
